crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,529 @@
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[allow(dead_code)]
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetcher: Sync + Send {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its native price
+    /// estimator result
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError>;
+}
+
+/// Buffered implementation that implements automatic batching of
+/// native prices requests.
+#[allow(dead_code)]
+#[derive(Clone)]
+pub struct BufferedTransport<Inner> {
+    config: Configuration,
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+/// Object to map the token with its native price estimator result
+#[allow(dead_code)]
+#[derive(Clone)]
+struct NativePriceResult {
+    token: H160,
+    result: Result<f64, PriceEstimationError>,
+}
+
+#[allow(dead_code)]
+impl<Inner> BufferedTransport<Inner>
+where
+    Inner: NativePriceBatchFetcher + Send + Sync + 'static,
+{
+    /// Maximum capacity of the broadcast channel, the messages are discarded as
+    /// soon as they are sent, so this limit should be enough to hold the
+    /// flow
+    const BROADCAST_CHANNEL_CAPACITY: usize = 50;
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+
+        let (broadcast_sender, _) = broadcast::channel(Self::BROADCAST_CHANNEL_CAPACITY);
+
+        Self::background_worker(
+            inner.clone(),
+            config.clone(),
+            receiver,
+            broadcast_sender.clone(),
+        );
+
+        Self {
+            inner,
+            calls,
+            broadcast_sender,
+            config,
+        }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<H160>,
+        broadcast_sender: broadcast::Sender<NativePriceResult>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if !batch.is_empty() {
+                    let results: Vec<_> = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| NativePriceResult {
+                                token,
+                                result: price,
```

Comment:
Well, it used to be like that but I changed it to address a comment, since the type `(H160, PriceEstimateResult)` may seem complex, but I am open to either option.